### DISCUSSION TITLE
Support for Android Auto

### DISCUSSION THREAD

**Discussion by SamuelBrucksch** [mapsforge/mapsforge#1214] [Q&A] [answered]
Hi,

i'm interested in using this lib for my android auto navigation app. We want to use vector maps which is not possible with our current renderer. I had a look at the MapView and found following code:
https://github.com/mapsforge/mapsforge/blob/b2c3393fd67fe96c068709c6970ca4b1d848ab6c/mapsforge-map-android/src/main/java/org/mapsforge/map/android/view/MapView.java#L313-L321

It looks like the framebuffer draws on the canvas, and that's it more or less behind for drawing. This is perfect, as it uses the same mechanism as for Android Auto. Please have a look at this AA sample code:
https://github.com/android/car-samples/blob/2a42b5488ef897e068365aa5557ac19ea2634ebb/car_app_library/navigation/src/main/java/androidx/car/app/samples/navigation/car/SurfaceRenderer.java#L157

We get a canvas from Android Auto where we can draw on.

So if i compare it to the code from the MapView, it would look like this for Android Auto:
```
Canvas canvas = mSurface.lockCanvas(null);
this.frameBuffer.draw(canvas);
if (this.mapScaleBar != null) {
    this.mapScaleBar.draw(canvas);
}
this.fpsCounter.draw(canvas);
mSurface.unlockCanvasAndPost(canvas);
```

Is there a way to use that logic behind, that does the drawing, without using the Android MapView?

Do you maybe have something like that somewhere in the code where you can point me to?

Thanks!

**Comment by devemux86** [comment]
What do you mean?

The library has a common graphics API to be applied on Android and Desktop.

Can't already use draw calls in your context or create your custom map view?

**Reply by SamuelBrucksch** [reply]
Is it as easy as copying the code from mapview and calling our own draw function? I could try that then... Some libs require special initialization or hide some things so you can't access them in different packages and so on.

**Comment by devemux86** [comment]
You need to test it.

[mapsforge-map-android](https://github.com/mapsforge/mapsforge/tree/master/mapsforge-map-android) module works fine on mobile,
but could need an extension module with different implementation for Android Auto.

**Reply by SamuelBrucksch** [reply]
I want to use vector maps, so i guess i'm in the wrong project, will open another discussion at VTM. Thanks

**Comment by devemux86** [comment]
Both map libraries, [Mapsforge](https://github.com/mapsforge/mapsforge) and [VTM](https://github.com/mapsforge/vtm), can read Mapsforge vector maps.

They have different map engines, VTM uses OpenGL.
And both use a common graphics API to work on multiple platforms.

**Reply by SamuelBrucksch** [reply]
Ah ok, good to know, i thought vector maps only will work with VTM. But that one is quite hard to understand and not so easy to integrate into AA. Guess I'll give it another try.

**Comment by SamuelBrucksch** [comment]
I can't seem to find documentation on how to get vector maps into a mapsforge map. Could you please point me to to some documentation / sample? I want to use an online vector tile source with z/x/y.pbf (mapbox) format.

Also i could not find any information about a 3D map mod / tilting of map, can i achieve that as well with mapsforge, or would VTM be a better choice here?

**Comment by devemux86** [comment, marked as answer]
Both support Mapsforge offline vector map files.

VTM with OpenGL supports online vector maps and has map rotation, tilt, etc.

There are getting started examples and many samples in their repositories.

**Reply by SamuelBrucksch** [reply]
Ok so i have to use VTM. Will open another discussion there then. Thanks for the help so far.
